include/net/zstream.h

```diff
@@ -0,0 +1,60 @@
+/**
+ * @file
+ * @brief Network stream API definitions
+ *
+ * An API to abstract different transport protocols for SOCK_STREAMs, etc.
+ */
+
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __NET_ZSTREAM_H
+#define __NET_ZSTREAM_H
+
+struct zstream_api;
+
+struct zstream_obj {
+	const struct zstream_api *api;
+};
+
+typedef struct zstream_obj *zstream;
```

Comment:
Having typedef for structs is not very zephyr like as we typically do not have that. And hiding the fact that zstream is actually a pointer is imho very confusing and bad practice.
Why just not have
```
struct zstream {
      const struct zstream_api *api
};
```
and get rid of zstream_obj and use "struct zstream *" instead of "zstream".

Reply:
@jukkar:

Well, struct is something which has (more or less public) fields. On the other hand, `zstream` is a type of object. It doesn't have any public internal structure, and the only way to operate on it is using its methods (which are functions taking `zstream` type as a first param in C). Indeed, we don't use such a pattern often in Zephyr, but sometimes we do, e.g. `k_tid_t` is an opaque type of an object representing thread, i.e. has the same usecase as `zstream`.

Anyway, that describes why it was done that way. I'm not against changing such "meta-interface" things (this also includes @tbursztyka's suggestion for use shorter namespacing prefixes), but let's put them off until the functionality is complete and "finalized" implementation is presented (I also wished there was more than 1 vote to change them, as IMHO "pro" arguments are good too, but up to persistence of the reviewers ;-) ).

Reply:
> use "struct zstream *" instead of "zstream".

Based on the discussion, this has now been implemented.